fix(api): Skip picking up and dropping tips for 0-volume steps in a transfer

alexjoel42 opened the pull request:
# Overview

Closes #8015.

Found a bug/differing behavior where during a transfer command, if the argument ` new_tip = "always"` is provided, then the pipettes would pick up and drop tips even if they weren't aspirating/dispensing anything during that step. Reasons to change this would be:

- Customers accidentally can waste tips when reading from a .csv file
- It is inconsistent with how we treat transfers since the API fix for #6620

# Changelog
- Fixed tip pickup and drop for 0 µL transfers while the argument `new _tip = "always"`  is there

To show how this behavior effects protocol runs, I ran a quick demo:

## Current Behavior 
This is where there will be a pick up and drop tip for 0 volume. 

<img width="1001" alt="Screen Shot 2021-09-03 at 5 19 00 PM" src="https://user-images.githubusercontent.com/65437763/132066238-0c42f3eb-1bed-4979-81a7-bf5a741dfb80.png">


## After PR 
This will fully ignore the 0 volume

<img width="1179" alt="Screen Shot 2021-09-03 at 5 17 05 PM" src="https://user-images.githubusercontent.com/65437763/132066105-e399dd5e-d167-421c-a4ba-e308aa4535f5.png">

# Review requests

1. I was able to run a few variations of protocols without the bug reoccurring 
2. Please either comment or reach out separately about any improvements I could make when submitting Pull Requests. 

# Risk assessment

Low, moving new tip behavior into a for loop. 

To give more detail:

1. This is a change local to [transfers.py](https://github.com/Opentrons/opentrons/blob/API-bug-fix-8015/api/src/opentrons/protocols/advanced_control/transfers.py)
2 Within that context it only effects transfer functions with the `new_tip = "always"` argument provided (defined in `def _plan_transfer(self)`) 
3. I am unclear in how it would work with fast-sim, but it works

alexjoel42 committed:
Update transfers.py

for some reason my Git branch isn't accepting pushes.

alexjoel42 committed:
Update transfers.py

for some reason my Git branch isn't accepting pushes. Hopefully this will update to the correct branch

alexjoel42 committed:
Update transfers.py

for some reason my Git branch isn't accepting pushes. Hopefully this will update to the correct branch

alexjoel42 committed:
Update transfers.py

alexjoel42 committed:
Merge remote-tracking branch 'origin/API-bug-fix-8015' into API-bug-fix-8015

alexjoel42 committed:
Update transfers.py

I accidentally didn't delete this in the PR

alexjoel42 committed:
Update transfers.py

I accidentally didn't delete this in the PR

sfoster1 commented on api/src/opentrons/protocols/advanced_control/transfers.py:
```diff
@@ -493,22 +493,23 @@ def _plan_transfer(self):
             xferred_vol = 0.0
             while xferred_vol < step_vol:
+                # only pick up/drop tips when the volume is greater than zero
+                if self._strategy.new_tip == types.TransferTipPolicy.ALWAYS:
```
If I'm understanding this correctly, this does fix the bug but it also introduces a new behavior (which might also be desired): If you're doing a transfer that requires more than one action per well (for instance, transferring 30 uL with a 20 uL tip, or transferring 20 uL + 5 uL airgap with a 20 uL tip), then each of the moves this is split into will pick up a new tip. If that's ok that's fine, but we should note it down.

sfoster1 replied:
If you don't want this behavior, probably the better thing to do is between line 495 and 496 do something like
```python
if step_vol < (some minimum value - don't do an exact equality comparison):
    continue
```
which should skip the rest of the body of the `for` loop on line 495.

alexjoel42 replied:
I think with regards to being loyal to the original intent of the PR (have 0 tip pickups and drop tips for zero volumes) I would like to follow the suggestion you proposed 

```python3
if step_vol < (some minimum value - don't do an exact equality comparison):
    continue
```

It might be worth re-examining at a later point to see if we'd like a different behavior  for transfers that require more than one action per well

alexjoel42 replied:
I made an attempt to fix it. but I think I accidentally ended up saying the same thing a different way. 
Fortunately, it does match with the expected behavior from our docs.

sfoster1 reviewed (changes requested):
The code looks good but especially given the follow-on behavior change we should probably alter the documentation [here](https://github.com/Opentrons/opentrons/blob/45f1c61a54256ef2b53ce486d67db13080408d55/api/docs/v2/new_complex_commands.rst#L582-L614). I think what's in the docs suggests this would be the behavior, but it hasn't been, so adding a note that says like "A new tip will be used even if multiple liquid handling steps are required to transfer each well" with an example runlog is probably a good idea.

alexjoel42 committed:
Update transfers.py

for some reason my Git branch isn't accepting pushes. Hopefully this will update to the correct branch

alexjoel42 committed:
Update transfers.py

for some reason my Git branch isn't accepting pushes. Hopefully this will update to the correct branch

SyntaxColoring reviewed (changes requested):
Awesome! Thanks for the PR.

-----

Strictly, I think we have to apply this fix conditionally, depending on the protocol's `apiLevel`.

If someone writes a protocol that relies on this fix, we want to force them to specify a new `apiLevel` so that outdated robots *without* this fix can provide a helpful error message.

In other words, we want to avoid this situation:

1. You have an up-to-date OT-2.
2. You write a protocol that implicitly relies on this fix. The protocol works on your OT-2.
3. You send the protocol to your colleague, who is on an older robot software version that doesn't have this fix.
4. When your colleague uploads this protocol to their OT-2, they get an "out of tips" error. Which is very confusing—it worked on your OT-2, and you haven't changed anything, so why isn't it working for your colleague?
    * Instead, your colleague should get an error that tells them to update their OT-2.

(Yes, this policy is burdensome. No, I'm not sure if it's the right policy for us to have. See https://github.com/Opentrons/opentrons/issues/7477#issuecomment-801414763 for some thinking that I haven't gotten around to copying into a wiki.)

If we agree that we should do this, I can help @alexjoel42 implement it.

alexjoel42 committed:
Adjusting for the API versions this way the behavior will only be applied for later API versioning

alexjoel42 committed:
Adjusting for the API versions this way the behavior will only be applied for later API versioning

alexjoel42 commented:
@SyntaxColoring I think that I fixed the #7477 ([comment](https://github.com/Opentrons/opentrons/issues/7477#issuecomment-801414763)) formatting request. 

https://github.com/Opentrons/opentrons/tree/API-bug-fix-8015

So that it only uses API versions 2.11 and above. 


<img width="872" alt="Screen Shot 2021-09-13 at 9 26 04 AM" src="https://user-images.githubusercontent.com/65437763/133091796-f34902ef-cc6f-401b-8b0e-c3f89975bf61.png">

sfoster1 commented on api/src/opentrons/protocols/advanced_control/transfers.py:
```diff
@@ -492,23 +492,45 @@ def _plan_transfer(self):
+                    vol = min(max_vol, step_vol - xferred_vol)
+                    yield from self._aspirate_actions(vol, src)
+                    yield from self._dispense_actions(vol=vol, dest=dest, src=src)
+                    yield from self._new_tip_action()
```
I don't think this is in the right place; I think it needs to be where it was, outside, the `while` loop, for this implementation of pick up tip. `self._new_tip_action` is `drop_tip`/`return_tip`, so it should be called once for every `pick_up_tip`.

alexjoel42 replied:
@sfoster1  I think I need another push on this one to get the test case ready. I'm still a bit stumped. 

I think that the current format follows the [API documentation ](https://docs.opentrons.com/v2/new_complex_commands.html#always-get-a-new-tip)'s description 
> The pipette can optionally get a new tip at the beginning of each aspirate, to help avoid cross contamination.

Do you have any recommendations for building a better test case? 

The test case that I used was 

```Python3

from opentrons import protocol_api, types
metadata = {'apiLevel': '2.12'}

def run(protocol: protocol_api.ProtocolContext):
    tr1 = protocol.load_labware('opentrons_96_tiprack_300ul', '3')
    p300 = protocol.load_instrument('p300_multi_gen2', 'right', tip_racks=[tr1])
    reservoir = protocol.load_labware('nest_12_reservoir_15ml', location='2')
    well = protocol.load_labware('nest_96_wellplate_2ml_deep', location='4')

    p300.transfer([1000,0], reservoir['A1'], [well['A3'], well['A2']], new_tip='Always')

```


##  The original Code that has the additional "pick_up and drop_tip with 0 volume"
<img width="846" alt="Screen Shot 2022-03-30 at 4 34 15 PM" src="https://user-images.githubusercontent.com/65437763/160926018-f480ed4e-0a1e-414b-8347-b4fe86ddc8e8.png">


## V2 The modified behavior 
<img width="852" alt="Screen Shot 2022-03-30 at 4 33 56 PM" src="https://user-images.githubusercontent.com/65437763/160925955-cb832cf2-f797-4343-9569-60b667572e3e.png">

alexjoel42 replied:
Fooey. I was wondering why it said "526 commits." I still need to rebase this

sfoster1 reviewed (changes requested):
I think there's another slight problem with the more-than-max-volume transfer case. I think a good thing to do here might be to step back and fix up the test cases as a next step; that may help you lock down the behavior you're looking for.

alexjoel42 committed:
I am unclear what the greater than max volume behavior is different. I thought they were the same

alexjoel42 committed:
I am unclear what the greater than max volume behavior is different. I thought they were the same

alexjoel42 committed:
Merge remote-tracking branch 'origin/API-bug-fix-8015' into API-bug-fix-8015

alexjoel42 committed:
we are good to go with API version and longer commands

alexjoel42 committed:
Preparing for rebase

alexjoel42 committed:
Merge remote-tracking branch 'origin/edge' into API-bug-fix-8015

mcous reviewed (commented):
Noticed some pushes recently on this branch, and figured we should actually try to make sure this gets in! Two notes:

- Since this PR was opened, PAPIv2.12 was released, so this PR would need to bump to 2.13
- We need tests included with this PR to verify the new behavior(s) exposed

alexjoel42 commented:
@mcous I would like to get this live, I am just having trouble with making the right test cases. As far as I'm aware, the normal behavior is unchanged. Here is what I have tried so far. 

# Volume higher than max 

```python3 
from opentrons import protocol_api, types
metadata = {'apiLevel': '2.12'}

def run(protocol: protocol_api.ProtocolContext):
    tr1 = protocol.load_labware('opentrons_96_tiprack_300ul', '3')
    p300 = protocol.load_instrument('p300_multi_gen2', 'right', tip_racks=[tr1])
    reservoir = protocol.load_labware('nest_12_reservoir_15ml', location='2')
    well = protocol.load_labware('nest_96_wellplate_2ml_deep', location='4')

    p300.transfer([1000,0], reservoir['A1'], [well['A3'], well['A2']], new_tip='Always')
```

## Behaves the same except for bug fix 

This is a side by side comparison of my version vs. the current version
![Screen Shot 2022-04-11 at 10 06 18 AM](https://user-images.githubusercontent.com/65437763/162757184-f134194a-f990-471b-8e86-449b1853d534.png)


# Normal transfer with the same values. 
```Python3 
from opentrons import protocol_api, types
metadata = {'apiLevel': '2.12'}

def run(protocol: protocol_api.ProtocolContext):
    tr1 = protocol.load_labware('opentrons_96_tiprack_300ul', '3')
    p300 = protocol.load_instrument('p300_multi_gen2', 'right', tip_racks=[tr1])
    reservoir = protocol.load_labware('nest_12_reservoir_15ml', location='2')
    well = protocol.load_labware('nest_96_wellplate_2ml_deep', location='4')

    p300.transfer([200,0], reservoir['A1'], [well['A3'], well['A2']], new_tip='Always')

```
## Behaves the same except for bug fix 

![Screen Shot 2022-04-11 at 10 09 30 AM](https://user-images.githubusercontent.com/65437763/162757790-d193ab8c-45d6-4261-8c91-ac2ca08c27ac.png)

## Never and Always both behave normally

So I'm a bit confused about what still needs to be done aside from 2.13

mcous commented:
@alexjoel42 my apologies for not being more clear. When I saw "we need tests included" I am referring to the automated unit tests in `api/tests/opentrons/protocols/advanced_control/test_transfers.py`.

This PR needs to include new tests in that file. In general, all PRs that add or change functionality much include accompanying unit tests

alexjoel42 commented:
Aha! 

That makes sense. I haven't used Pytest before, exciting to learn a new skill. 

Honestly, I am unsure what the error means? Especially since I was able to run the code with blow outs. 


 
![Screen Shot 2022-04-11 at 11 04 37 AM](https://user-images.githubusercontent.com/65437763/162769030-ca9140ca-cd52-4f79-888b-b094d0218512.png)

laviera commented:
Closing PR - to be picked up by RS if needed post release

mcous commented:
@alexjoel42 feel free to continue pushing to this branch if/when you have time! We can re-open the PR if the changes discussed above get made. In the mean time we just didn't want a relatively inactive PR crowding up the "needs review" page, hence the close
